code/modules/projectiles/projectile.dm

```diff
@@ -170,15 +170,28 @@
 			permutated.Add(A)
 		return FALSE
 	else
-		if(A && A.density && !ismob(A) && !(A.flags_1 & ON_BORDER_1)) //if we hit a dense non-border obj or dense turf then we also hit one of the mobs on that tile.
+		if(A && A.density && !ismob(A) && !istype(A,/obj/machinery) && !istype(A,/obj/structure/) && !(A.flags_1 & ON_BORDER_1)) //if we hit a dense non-border obj or dense turf then we also hit one of the mobs on that tile.
 			var/list/mobs_list = list()
+			var/list/machine_list = list()
 			for(var/mob/living/L in target_turf)
 				mobs_list += L
-			if(mobs_list.len)
-				var/mob/living/picked_mob = pick(mobs_list)
-				if(!prehit(picked_mob))
-					return FALSE
-				picked_mob.bullet_act(src, def_zone)
+			for(var/obj/machinery/m in target_turf)
+				machine_list += m
+			if(mobs_list.len || machine_list.len)
+				if(mobs_list.Find(original) || machine_list.Find(original))
+					if(!prehit(original))

```

Comment:
Instead of duplicating the 
```DM
if(!prehit(original))
	return FALSE
thing.bullet_act(src, def_zone)
```
code three times,make a ```var/atom/movable/hit_thing```, set hit_thing in this if-else chain, then after the if-else chain just run this code on hit_thing.